### PR TITLE
Include directory entries when building wheel

### PR DESCRIPTION
**Pull request by jaraco**
See #287 for background.

### CODE REVIEW

File: wheel/wheelfile.py [hunk 1]
```diff
@@ -136,12 +142,19 @@ def write(self, filename, arcname=None, compress_type=None):
...
+        zinfo.compress_type = ZIP_DEFLATED
+        self.writestr(zinfo, b'')
+
     def writestr(self, zinfo_or_arcname, bytes, compress_type=None):
         ZipFile.writestr(self, zinfo_or_arcname, bytes, compress_type)
         fname = (zinfo_or_arcname.filename if isinstance(zinfo_or_arcname, ZipInfo)
                  else zinfo_or_arcname)
         logger.info("adding '%s'", fname)
-        if fname != self.record_path:
+        if fname != self.record_path and not fname.endswith('/'):
```
Comment (agronholm) [inline wheel/wheelfile.py]: Is this check really necessary?

Comment (jaraco) [reply wheel/wheelfile.py]: If you take it out, the tests fail because there are RECORD entries for the directories (with a SHA hash for an empty file).

Comment (agronholm) [reply wheel/wheelfile.py]: Okay, I see it's also used by the `mkdir()` method.

File: tests/test_wheelfile.py [hunk 2]
```diff
@@ -172,3 +173,19 @@ def test_attributes(tmpdir_factory, wheel_path):
         info = zf.getinfo('test-1.0.dist-info/RECORD')
         permissions = (info.external_attr >> 16) & 0o777
         assert permissions == 0o664
+
+
+def test_directories(tmpdir, wheel_path):
+    build_dir = tmpdir
+    sub_dir = build_dir / 'sub'
+    sub_dir.mkdir()
+    (sub_dir / '__init__.py').write_text('', encoding='utf-8')
```
Comment (agronholm) [inline tests/test_wheelfile.py]: Why create a file in the directory? Wasn't the point to ensure that *empty* directories get included, or did I misunderstand?

Comment (jaraco) [reply tests/test_wheelfile.py]: Good question. I decided that it's probably better to match the behavior observed in other tools that generate zip files from a tree and create directory entries for every directory. But I do take from your point that the test should probably also ensure that directory entries are created for empty directories.

Comment (jaraco) [reply tests/test_wheelfile.py]: Updated in 0e94b10

**Comment by jaraco** [comment]
Woo! Thanks.

**Pull request merged**